pkg/controllers/workapplier/preprocess.go

```diff
@@ -77,13 +77,37 @@ func (r *Reconciler) preProcessManifests(
 
 		bundle.manifestObj = manifestObj
 		bundle.gvr = gvr
+
+		// Add the string representation of the work resource identifier to the bundle.
+		//
+		// Note that the string representation ignores the ordinal number in the identifier.
+		wriStr, err := formatWRIString(bundle.id)
+		if err != nil {
+			// The work resource identifier cannot be formatted as a string. The formatting will
+			// fail if and only if the manifest data cannot be decoded (a failure in the previous
+			// step); normally this branch will never run as such error should have already terminated
+			// the pre-processing stage.
+			klog.ErrorS(err, "Failed to format the work resource identifier as a string", "ordinal", pieces, "work", klog.KObj(work))
+			_ = controller.NewUnexpectedBehaviorError(fmt.Errorf("failed to format the work resource identifier as a string: %w", err))
+			return
+		}
+		bundle.wriStr = wriStr
+
 		klog.V(2).InfoS("Decoded a manifest",
 			"manifestObj", klog.KObj(manifestObj),
 			"GVR", *gvr,
 			"work", klog.KObj(work))
 	}
 	r.parallelizer.ParallelizeUntil(childCtx, len(bundles), doWork, "decodingManifests")
 
+	// Check for duplicated manifests.
+	//
+	// This is to address a corner case where users might have specified the same manifest
+	// twice in resource envelopes (or both in the envelopes and directly in the hub cluster).
+	//
+	// Note that the CRP/RP APIs will block repeated resource selectors.
+	checkForDuplicatedManifests(bundles, work)
```

Comment:
what's the diff between old and existing code? looks like the existing code already checked the duplicate manifests.